enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,773 @@
+- This feature does not aim to modify the Kubernetes scheduler
+- This feature does not "schedule" pods; it influences scheduling by patching
+  the affinity of gated pods[^2]
+- Modification of fields other than the node affinity is not supported.
+  Moreover, the controller logic will only consider the
+  `kubernetes.io/arch` label
+
+## Proposal
+Kubernetes 1.27 introduced the scheduling readiness feature[^1]. It
+allows users and controllers to set scheduling gates in the
+`podSpec.schedulingGates` list at admission time. Gated pods allow mutating some
+of the pod spec fields before the pod is scheduled. In particular, the
+scheduler's default `Enqueue` plugin holds the pods before entering the
+`ActiveQ` queue if a scheduling gate exists in the `podSpec.schedulingGates`
+list. Until gates are in the list, the pod will expose its status as
+`SchedulingGated`. The controller responsible for a given scheduling gate should
+implement the business logic so that the pod can be considered ready to be
+scheduled. KEP-3838[^2] introduced the possibility of mutating the node affinity
+of pods while they are in `schedulingGated` status (i.e., their
+`podSpec.schedulingGates` list has at least one gate).
+
+The multiarch-manager-operator will leverage the scheduling readiness feature
+and the mutable pod scheduling directives to update the node affinity of pods
+with the set of architectures they support. When a pod is submitted to the API
+server, a mutating admission webhook intercepts the pod and adds a scheduling
+gate to the pod spec.
+
+The `PodPlacementController` watches the pod resource. Suppose any pod has the
+scheduling gate set. In that case, the images in the pod spec are inspected, a
+subset of architectures that satisfy all containers in the pod is built, and the
+pod spec is patched with the additional node affinity constraints. If any
+existing node selector requirements are set with key `kubernetes.io/arch`,
+the `PodPlacementController` does not override the behavior previously
+pushed by other stakeholders, leaving it as complies with KEP-3838[^2].
+
+The multiarch-manager-operator will be available as part of the OperatorHub, as
+a secondary operator managed by the Operator Lifecycle Manager (OLM). It will be
+a cluster-scoped operator that only cluster admins can install and configure.
+
+Its development is expected to ship in the following phases:
+
+- Phase 1 - Technology Preview (TP):
+  - The pod placement controller will be supported in High Availability (HA)
+    mode
+  - The webhook will be a stateless component
+  - The pod placement controller will support namespace filtering to skip pods
+    that should not be mutated
+  - No support for customized network scenarios (restricted, proxied,
+    disconnected...)
+  - No TLS verification for the registries as we are only fetching metadata
+  - No caching facilities for the image metadata
+  - No graceful uninstallation support: any remaining gated pods have to be
```

Comment:
Yes,  as described in the risks and mitigations section, the pod would need to be manually patched to remove the scheduling gate to continue with normal scheduling. We might tackle this in later phases - the problem here is that when the operator is uninstalled OLM does not have a concept of ordered deletion. So it would remove the RBAC rules needed to access the pods and the controller would then fail when it tries to get a list of pods to remove the scheduling gate before exiting.

Reply:
@joelanford do you know if there is a way that OLM operators do this sort of cleanup?

Reply:
Oh right, yeah, sounds like you need some way to uninstall things, but that would probably mean somehow using finalizers to make sure you have all the rbac/resources required to clean up, and, well, it's going to get messy isn't it 🤔

Reply:
that is correct. even if we use finalizers on the controller - the rbac rules will be cleaned up. not sure if there is a way to set finalizers for rbac and other resources.

Reply:
You should be able to set a finalizer on any object, the field is present in the API for every object in Kube so I'd expect it to work

Reply:
we will have it documented and yes, should be simple enough to provide a script. i updated the risks and mitigations section with this detail.

Reply:
@Prashanth684 can this conversation be resolved?